### PR TITLE
Sync code changes from darktable

### PR DESCRIPTION
**Pull request by pedrocr**
This includes support for new RAF formats and then a bunch of safety fixes. afl-fuzz and AddressSanitizer are much happier now, but there is probably still a bit more to fix. This also includes a partial implementation of the ideas in issue #142. My next step will be to remove TiffEntryBE/TiffIFDBE entirely.

### CODE REVIEW

File: RawSpeed/RawImage.cpp [hunk 1]
```diff
@@ -326,7 +326,9 @@ void RawImageData::startWorker(RawImageWorker::RawImageWorkerTask task, bool cro
 void RawImageData::fixBadPixelsThread( int start_y, int end_y )
 {
   int gw = (uncropped_dim.x + 15) / 32;
+#ifdef __AFL_COMPILER
```
Comment (klauspost) [inline RawSpeed/RawImage.cpp:329]: What is this for?

File: RawSpeed/RawImage.cpp [hunk 2]
```diff
@@ -337,8 +339,10 @@ void RawImageData::fixBadPixelsThread( int start_y, int end_y )
         for (int i = 0; i < 4; i++) {
           for (int j = 0; j < 8; j++) {
             if (1 == ((bad[i]>>j) & 1)) {
-              if (bad_count++ > 500)
+#ifdef __AFL_COMPILER
+              if (bad_count++ > 100)
```
Comment (klauspost) [inline RawSpeed/RawImage.cpp:343]: Why this change?

File: RawSpeed/RawImage.cpp (hunk 1 above)
Comment (pedrocr) [inline RawSpeed/RawImage.cpp:329]: afl-fuzz works best when the program is very fast, so it can try many samples per second. It's easy to generate a pathological file that has so many bad pixels the run-time goes quadratic. So when compiling under AFL just limit the bad pixels to 100 and bail out after that. It's enough for the code to be tested but doesn't make the fuzzing too slow.

I tried just having a higher limit enabled all the time but the zero_is_bad stuff in RW2 means they actually have a lot of bad pixels detected. Might be a bug in RW2 though.

File: RawSpeed/RawImage.cpp (hunk 2 above)
Comment (pedrocr) [inline RawSpeed/RawImage.cpp:343]: See the other comment, with RW2 there's no sensible limit that will work so I just enabled it only when running under afl and then made the limit smaller to be more effective.

**Comment by klauspost** [comment]
It definitely makes the code much nicer on a quick look-through.

**Comment by pedrocr** [comment]
:) cool. I've done regression testing against all the files in rawsamples.ch and all of them are unchanged, so hopefully it's correct too. I pushed it to darktable master yesterday so our more adventurous users can find any potential weird bugs, but the regression testing coverage is usually pretty good at catching everything.

File: RawSpeed/TiffEntryBE.cpp [hunk 3]
```diff
@@ -84,67 +81,77 @@ TiffEntryBE::TiffEntryBE( TiffTag tag, TiffDataType type, uint32 count, const uc
...
+    if (b)
+      return (float) a/b;
+  } else if (type == TIFF_SRATIONAL) {
+    int a = (int) getInt(num*2);
+    int b = (int) getInt(num*2+1);
+    if (b)
+      return (float) a/b;
   }
-  return (short16 *)d;
+  return 0.0f;
```
Comment (klauspost) [inline RawSpeed/TiffEntryBE.cpp]: AFAICT this return "0", if the type isn't known.

Shouldn't there be an "else throw error" in here?

**Comment by klauspost** [comment]
Cheers. It seems to almost only affect meta-data reading, so I this the cleaner code appears to be worth it. 

Nice job!
